core/src/main/java/net/akami/yggdrasil/item/SpellTriggerItem.java

```diff
@@ -1,102 +1,105 @@
+import org.spongepowered.api.item.enchantment.EnchantmentTypes;
+import org.spongepowered.api.item.inventory.ItemStack;
+import org.spongepowered.api.text.Text;
+import org.spongepowered.api.text.format.TextColors;
+import org.spongepowered.api.world.World;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+public class SpellTriggerItem extends InteractiveAimingItem {
+
+    private ItemStack item;
+    private MagicUser user;
+
+    public SpellTriggerItem(MagicUser user) {
+        super(user);
+        this.user = user;
+        List<Enchantment> enchantments = Collections.singletonList(
+                Enchantment.of(EnchantmentTypes.INFINITY, 1));
+        Text name = Text.builder()
+                .color(TextColors.GOLD)
+                .append(Text.of("Launch Spell"))
+                .build();
+        this.item = ItemStack
+                .builder()
+                .itemType(ItemTypes.BOW)
+                .add(Keys.DISPLAY_NAME, name)
+                .add(Keys.UNBREAKABLE, true)
+                .add(Keys.ITEM_ENCHANTMENTS, enchantments)
+                .add(Keys.HIDE_ENCHANTMENTS, true)
+                .add(Keys.HIDE_ATTRIBUTES, true)
+                .add(Keys.HIDE_UNBREAKABLE, true)
+                .quantity(1)
+                .build();
+    }
+
+    @Override
+    protected void applyEffect(Vector3d location, World world) {
+        Optional<SpellCastContext> optResult = user.findBySequence();
+        user.clearSequence();
+        optResult.ifPresent(result -> applyEffect(location, result));
+    }
+
+    @Override
+    public void onRightClicked(CancellableEvent<?> event, GameItemClock clock) {
+        if(!aimlessSpell(event)) {
+            super.onRightClicked(event, clock);
+        }
+    }
+
+    private boolean aimlessSpell(Cancellable event) {
+
+        if(!ready) {
+            return false;
+        }
+        Optional<SpellCastContext> optResult = user.findBySequence();
+
+        if(optResult.isPresent()) {
+            SpellCastContext result = optResult.get();
+            if(!result.requiresLocation()) {
+                user.clearSequence();
+                applyEffect(null, result);
+                event.setCancelled(true);
+                ready = true;
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private void applyEffect(Vector3d location, SpellCastContext context) {
+        SpellCaster caster = context.getCaster();
+        int tier = context.getChosenTier();
+        castSpell(caster, location, tier);
+    }
+
+    private void castSpell(SpellCaster caster, Vector3d location, int tier) {
+        user.getMana().ifEnoughMana(caster.getCastingCost(tier), () -> {
+            Spell spell = caster.createSpell();
+            spell.cast(user, location, tier);
+        });
+    }
+
+    @Override
+    public ItemStack matchingItem() {
+        return item;
+    }
+}
```

Comment:
what happened to that file 🤔 nothing was changed yet the whole original file is in red, replaced by the exact same code in green, except for 3 imports (not that it really matters erm)

Reply:
Here, I don't know what happenned. I just changed, 2 parameters and 3 imports.